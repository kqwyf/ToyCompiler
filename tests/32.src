int[10][16] a;
float[20] b;

int main() {
    int m, n;
    m = 1;
    n = 2;
    bool[5] c;
    a[0][1] = 1;
    b[2] = 2;
    b[a[0][1]] = b[2];
    if(c[m + n])
        m = 1;
    else if(c[m] && c[n])
        n = 2;
}
