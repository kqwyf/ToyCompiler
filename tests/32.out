Token sequence:
int             < INT         ,        >
[               < LEFTBRACKET ,        >
10              < CONSTANT    , 1      >
]               < RIGHTBRACKET,        >
[               < LEFTBRACKET ,        >
16              < CONSTANT    , 2      >
]               < RIGHTBRACKET,        >
a               < IDENTIFIER  , 3      >
;               < SEMICOLON   ,        >
float           < FLOAT       ,        >
[               < LEFTBRACKET ,        >
20              < CONSTANT    , 4      >
]               < RIGHTBRACKET,        >
b               < IDENTIFIER  , 5      >
;               < SEMICOLON   ,        >
int             < INT         ,        >
main            < IDENTIFIER  , 6      >
(               < LEFTPAREN   ,        >
)               < RIGHTPAREN  ,        >
{               < LEFTBRACE   ,        >
int             < INT         ,        >
m               < IDENTIFIER  , 7      >
,               < COMMA       ,        >
n               < IDENTIFIER  , 8      >
;               < SEMICOLON   ,        >
m               < IDENTIFIER  , 7      >
=               < ASSIGN      ,        >
1               < CONSTANT    , 9      >
;               < SEMICOLON   ,        >
n               < IDENTIFIER  , 8      >
=               < ASSIGN      ,        >
2               < CONSTANT    , 10     >
;               < SEMICOLON   ,        >
bool            < BOOL        ,        >
[               < LEFTBRACKET ,        >
5               < CONSTANT    , 11     >
]               < RIGHTBRACKET,        >
c               < IDENTIFIER  , 12     >
;               < SEMICOLON   ,        >
a               < IDENTIFIER  , 3      >
[               < LEFTBRACKET ,        >
0               < CONSTANT    , 13     >
]               < RIGHTBRACKET,        >
[               < LEFTBRACKET ,        >
1               < CONSTANT    , 9      >
]               < RIGHTBRACKET,        >
=               < ASSIGN      ,        >
1               < CONSTANT    , 9      >
;               < SEMICOLON   ,        >
b               < IDENTIFIER  , 5      >
[               < LEFTBRACKET ,        >
2               < CONSTANT    , 10     >
]               < RIGHTBRACKET,        >
=               < ASSIGN      ,        >
2               < CONSTANT    , 10     >
;               < SEMICOLON   ,        >
b               < IDENTIFIER  , 5      >
[               < LEFTBRACKET ,        >
a               < IDENTIFIER  , 3      >
[               < LEFTBRACKET ,        >
0               < CONSTANT    , 13     >
]               < RIGHTBRACKET,        >
[               < LEFTBRACKET ,        >
1               < CONSTANT    , 9      >
]               < RIGHTBRACKET,        >
]               < RIGHTBRACKET,        >
=               < ASSIGN      ,        >
b               < IDENTIFIER  , 5      >
[               < LEFTBRACKET ,        >
2               < CONSTANT    , 10     >
]               < RIGHTBRACKET,        >
;               < SEMICOLON   ,        >
if              < IF          ,        >
(               < LEFTPAREN   ,        >
c               < IDENTIFIER  , 12     >
[               < LEFTBRACKET ,        >
m               < IDENTIFIER  , 7      >
+               < PLUS        ,        >
n               < IDENTIFIER  , 8      >
]               < RIGHTBRACKET,        >
)               < RIGHTPAREN  ,        >
m               < IDENTIFIER  , 7      >
=               < ASSIGN      ,        >
1               < CONSTANT    , 9      >
;               < SEMICOLON   ,        >
else            < ELSE        ,        >
if              < IF          ,        >
(               < LEFTPAREN   ,        >
c               < IDENTIFIER  , 12     >
[               < LEFTBRACKET ,        >
m               < IDENTIFIER  , 7      >
]               < RIGHTBRACKET,        >
&&              < AND         ,        >
c               < IDENTIFIER  , 12     >
[               < LEFTBRACKET ,        >
n               < IDENTIFIER  , 8      >
]               < RIGHTBRACKET,        >
)               < RIGHTPAREN  ,        >
n               < IDENTIFIER  , 8      >
=               < ASSIGN      ,        >
2               < CONSTANT    , 10     >
;               < SEMICOLON   ,        >
}               < RIGHTBRACE  ,        >

Symbol table:
1     10
2     16
3     a
4     20
5     b
6     main
7     m
8     n
9     1
10    2
11    5
12    c
13    0

Production sequence:
TYPE_BASIC -> int
TYPE_ARRAY -> TYPE_BASIC [ constant ]
TYPE_ARRAY -> TYPE_ARRAY [ constant ]
TYPE -> TYPE_ARRAY
IDENTIFIER_S -> identifier
DECLARE_VAR -> TYPE IDENTIFIER_S ;
DECLARE -> DECLARE_VAR
DECLARE_S -> DECLARE
TYPE_BASIC -> float
TYPE_ARRAY -> TYPE_BASIC [ constant ]
TYPE -> TYPE_ARRAY
IDENTIFIER_S -> identifier
DECLARE_VAR -> TYPE IDENTIFIER_S ;
DECLARE -> DECLARE_VAR
DECLARE_S -> DECLARE_S DECLARE
TYPE_BASIC -> int
TYPE -> TYPE_BASIC
DECLARE_FUNC_BEGIN -> TYPE identifier (
DECLARE_FUNC_SIGN -> DECLARE_FUNC_BEGIN ) {
TYPE_BASIC -> int
TYPE -> TYPE_BASIC
IDENTIFIER_S -> identifier
IDENTIFIER_S -> IDENTIFIER_S , identifier
DECLARE_VAR -> TYPE IDENTIFIER_S ;
STATEMENT -> DECLARE_VAR
STATEMENT_S -> STATEMENT
EXPRESSION2 -> identifier
EXPRESSION3 -> EXPRESSION2
EXPRESSION4 -> EXPRESSION3
EXPRESSION5 -> EXPRESSION4
EXPRESSION6 -> EXPRESSION5
EXPRESSION7 -> EXPRESSION6
EXPRESSION2 -> constant
EXPRESSION3 -> EXPRESSION2
EXPRESSION4 -> EXPRESSION3
EXPRESSION5 -> EXPRESSION4
EXPRESSION6 -> EXPRESSION5
EXPRESSION7 -> EXPRESSION6
EXPRESSION8 -> EXPRESSION7
EXPRESSION8 -> EXPRESSION7 = EXPRESSION8
EXPRESSION -> EXPRESSION8
STATEMENT -> EXPRESSION ;
STATEMENT_S -> STATEMENT_S STATEMENT
EXPRESSION2 -> identifier
EXPRESSION3 -> EXPRESSION2
EXPRESSION4 -> EXPRESSION3
EXPRESSION5 -> EXPRESSION4
EXPRESSION6 -> EXPRESSION5
EXPRESSION7 -> EXPRESSION6
EXPRESSION2 -> constant
EXPRESSION3 -> EXPRESSION2
EXPRESSION4 -> EXPRESSION3
EXPRESSION5 -> EXPRESSION4
EXPRESSION6 -> EXPRESSION5
EXPRESSION7 -> EXPRESSION6
EXPRESSION8 -> EXPRESSION7
EXPRESSION8 -> EXPRESSION7 = EXPRESSION8
EXPRESSION -> EXPRESSION8
STATEMENT -> EXPRESSION ;
STATEMENT_S -> STATEMENT_S STATEMENT
TYPE_BASIC -> bool
TYPE_ARRAY -> TYPE_BASIC [ constant ]
TYPE -> TYPE_ARRAY
IDENTIFIER_S -> identifier
DECLARE_VAR -> TYPE IDENTIFIER_S ;
STATEMENT -> DECLARE_VAR
STATEMENT_S -> STATEMENT_S STATEMENT
EXPRESSION2 -> identifier
EXPRESSION3 -> EXPRESSION2
EXPRESSION2 -> constant
EXPRESSION3 -> EXPRESSION2
EXPRESSION4 -> EXPRESSION3
EXPRESSION5 -> EXPRESSION4
EXPRESSION6 -> EXPRESSION5
EXPRESSION7 -> EXPRESSION6
EXPRESSION8 -> EXPRESSION7
EXPRESSION -> EXPRESSION8
EXPRESSION3 -> EXPRESSION3 [ EXPRESSION ]
EXPRESSION2 -> constant
EXPRESSION3 -> EXPRESSION2
EXPRESSION4 -> EXPRESSION3
EXPRESSION5 -> EXPRESSION4
EXPRESSION6 -> EXPRESSION5
EXPRESSION7 -> EXPRESSION6
EXPRESSION8 -> EXPRESSION7
EXPRESSION -> EXPRESSION8
EXPRESSION3 -> EXPRESSION3 [ EXPRESSION ]
EXPRESSION4 -> EXPRESSION3
EXPRESSION5 -> EXPRESSION4
EXPRESSION6 -> EXPRESSION5
EXPRESSION7 -> EXPRESSION6
EXPRESSION2 -> constant
EXPRESSION3 -> EXPRESSION2
EXPRESSION4 -> EXPRESSION3
EXPRESSION5 -> EXPRESSION4
EXPRESSION6 -> EXPRESSION5
EXPRESSION7 -> EXPRESSION6
EXPRESSION8 -> EXPRESSION7
EXPRESSION8 -> EXPRESSION7 = EXPRESSION8
EXPRESSION -> EXPRESSION8
STATEMENT -> EXPRESSION ;
STATEMENT_S -> STATEMENT_S STATEMENT
EXPRESSION2 -> identifier
EXPRESSION3 -> EXPRESSION2
EXPRESSION2 -> constant
EXPRESSION3 -> EXPRESSION2
EXPRESSION4 -> EXPRESSION3
EXPRESSION5 -> EXPRESSION4
EXPRESSION6 -> EXPRESSION5
EXPRESSION7 -> EXPRESSION6
EXPRESSION8 -> EXPRESSION7
EXPRESSION -> EXPRESSION8
EXPRESSION3 -> EXPRESSION3 [ EXPRESSION ]
EXPRESSION4 -> EXPRESSION3
EXPRESSION5 -> EXPRESSION4
EXPRESSION6 -> EXPRESSION5
EXPRESSION7 -> EXPRESSION6
EXPRESSION2 -> constant
EXPRESSION3 -> EXPRESSION2
EXPRESSION4 -> EXPRESSION3
EXPRESSION5 -> EXPRESSION4
EXPRESSION6 -> EXPRESSION5
EXPRESSION7 -> EXPRESSION6
EXPRESSION8 -> EXPRESSION7
EXPRESSION8 -> EXPRESSION7 = EXPRESSION8
EXPRESSION -> EXPRESSION8
STATEMENT -> EXPRESSION ;
STATEMENT_S -> STATEMENT_S STATEMENT
EXPRESSION2 -> identifier
EXPRESSION3 -> EXPRESSION2
EXPRESSION2 -> identifier
EXPRESSION3 -> EXPRESSION2
EXPRESSION2 -> constant
EXPRESSION3 -> EXPRESSION2
EXPRESSION4 -> EXPRESSION3
EXPRESSION5 -> EXPRESSION4
EXPRESSION6 -> EXPRESSION5
EXPRESSION7 -> EXPRESSION6
EXPRESSION8 -> EXPRESSION7
EXPRESSION -> EXPRESSION8
EXPRESSION3 -> EXPRESSION3 [ EXPRESSION ]
EXPRESSION2 -> constant
EXPRESSION3 -> EXPRESSION2
EXPRESSION4 -> EXPRESSION3
EXPRESSION5 -> EXPRESSION4
EXPRESSION6 -> EXPRESSION5
EXPRESSION7 -> EXPRESSION6
EXPRESSION8 -> EXPRESSION7
EXPRESSION -> EXPRESSION8
EXPRESSION3 -> EXPRESSION3 [ EXPRESSION ]
EXPRESSION4 -> EXPRESSION3
EXPRESSION5 -> EXPRESSION4
EXPRESSION6 -> EXPRESSION5
EXPRESSION7 -> EXPRESSION6
EXPRESSION8 -> EXPRESSION7
EXPRESSION -> EXPRESSION8
EXPRESSION3 -> EXPRESSION3 [ EXPRESSION ]
EXPRESSION4 -> EXPRESSION3
EXPRESSION5 -> EXPRESSION4
EXPRESSION6 -> EXPRESSION5
EXPRESSION7 -> EXPRESSION6
EXPRESSION2 -> identifier
EXPRESSION3 -> EXPRESSION2
EXPRESSION2 -> constant
EXPRESSION3 -> EXPRESSION2
EXPRESSION4 -> EXPRESSION3
EXPRESSION5 -> EXPRESSION4
EXPRESSION6 -> EXPRESSION5
EXPRESSION7 -> EXPRESSION6
EXPRESSION8 -> EXPRESSION7
EXPRESSION -> EXPRESSION8
EXPRESSION3 -> EXPRESSION3 [ EXPRESSION ]
EXPRESSION4 -> EXPRESSION3
EXPRESSION5 -> EXPRESSION4
EXPRESSION6 -> EXPRESSION5
EXPRESSION7 -> EXPRESSION6
EXPRESSION8 -> EXPRESSION7
EXPRESSION8 -> EXPRESSION7 = EXPRESSION8
EXPRESSION -> EXPRESSION8
STATEMENT -> EXPRESSION ;
STATEMENT_S -> STATEMENT_S STATEMENT
EXPRESSION2 -> identifier
EXPRESSION3 -> EXPRESSION2
EXPRESSION2 -> identifier
EXPRESSION3 -> EXPRESSION2
EXPRESSION4 -> EXPRESSION3
EXPRESSION5 -> EXPRESSION4
EXPRESSION2 -> identifier
EXPRESSION3 -> EXPRESSION2
EXPRESSION4 -> EXPRESSION3
EXPRESSION5 -> EXPRESSION5 + EXPRESSION4
EXPRESSION6 -> EXPRESSION5
EXPRESSION7 -> EXPRESSION6
EXPRESSION8 -> EXPRESSION7
EXPRESSION -> EXPRESSION8
EXPRESSION3 -> EXPRESSION3 [ EXPRESSION ]
EXPRESSION4 -> EXPRESSION3
EXPRESSION5 -> EXPRESSION4
EXPRESSION6 -> EXPRESSION5
EXPRESSION7 -> EXPRESSION6
EXPRESSION8 -> EXPRESSION7
EXPRESSION -> EXPRESSION8
SELECT_BEGIN -> if ( EXPRESSION )
EXPRESSION2 -> identifier
EXPRESSION3 -> EXPRESSION2
EXPRESSION4 -> EXPRESSION3
EXPRESSION5 -> EXPRESSION4
EXPRESSION6 -> EXPRESSION5
EXPRESSION7 -> EXPRESSION6
EXPRESSION2 -> constant
EXPRESSION3 -> EXPRESSION2
EXPRESSION4 -> EXPRESSION3
EXPRESSION5 -> EXPRESSION4
EXPRESSION6 -> EXPRESSION5
EXPRESSION7 -> EXPRESSION6
EXPRESSION8 -> EXPRESSION7
EXPRESSION8 -> EXPRESSION7 = EXPRESSION8
EXPRESSION -> EXPRESSION8
STATEMENT -> EXPRESSION ;
SELECT_MID -> else
EXPRESSION2 -> identifier
EXPRESSION3 -> EXPRESSION2
EXPRESSION2 -> identifier
EXPRESSION3 -> EXPRESSION2
EXPRESSION4 -> EXPRESSION3
EXPRESSION5 -> EXPRESSION4
EXPRESSION6 -> EXPRESSION5
EXPRESSION7 -> EXPRESSION6
EXPRESSION8 -> EXPRESSION7
EXPRESSION -> EXPRESSION8
EXPRESSION3 -> EXPRESSION3 [ EXPRESSION ]
EXPRESSION4 -> EXPRESSION3
EXPRESSION5 -> EXPRESSION4
EXPRESSION6 -> EXPRESSION5
EXPRESSION7 -> EXPRESSION6
EXPRESSION2 -> identifier
EXPRESSION3 -> EXPRESSION2
EXPRESSION2 -> identifier
EXPRESSION3 -> EXPRESSION2
EXPRESSION4 -> EXPRESSION3
EXPRESSION5 -> EXPRESSION4
EXPRESSION6 -> EXPRESSION5
EXPRESSION7 -> EXPRESSION6
EXPRESSION8 -> EXPRESSION7
EXPRESSION -> EXPRESSION8
EXPRESSION3 -> EXPRESSION3 [ EXPRESSION ]
EXPRESSION4 -> EXPRESSION3
EXPRESSION5 -> EXPRESSION4
EXPRESSION6 -> EXPRESSION5
EXPRESSION7 -> EXPRESSION7 && EXPRESSION6
EXPRESSION8 -> EXPRESSION7
EXPRESSION -> EXPRESSION8
SELECT_BEGIN -> if ( EXPRESSION )
EXPRESSION2 -> identifier
EXPRESSION3 -> EXPRESSION2
EXPRESSION4 -> EXPRESSION3
EXPRESSION5 -> EXPRESSION4
EXPRESSION6 -> EXPRESSION5
EXPRESSION7 -> EXPRESSION6
EXPRESSION2 -> constant
EXPRESSION3 -> EXPRESSION2
EXPRESSION4 -> EXPRESSION3
EXPRESSION5 -> EXPRESSION4
EXPRESSION6 -> EXPRESSION5
EXPRESSION7 -> EXPRESSION6
EXPRESSION8 -> EXPRESSION7
EXPRESSION8 -> EXPRESSION7 = EXPRESSION8
EXPRESSION -> EXPRESSION8
STATEMENT -> EXPRESSION ;
SELECT -> SELECT_BEGIN STATEMENT
STATEMENT -> SELECT
SELECT -> SELECT_BEGIN STATEMENT SELECT_MID STATEMENT
STATEMENT -> SELECT
STATEMENT_S -> STATEMENT_S STATEMENT
DECLARE_FUNC -> DECLARE_FUNC_SIGN STATEMENT_S }
DECLARE -> DECLARE_FUNC
DECLARE_S -> DECLARE_S DECLARE
PROGRAM -> DECLARE_S

Semantic Symbol Tables:

Global Symbol Table (Table 0):
  #   |   DataType   |     Name     |  Offset  |    Attr
------+--------------+--------------+----------+------------
    0 | INT          | 10           | 0        | 
    1 | INT          | 16           | 4        | 
    2 | INT          | 20           | 8        | 
    3 | INT          | 1            | 12       | 
    4 | INT          | 2            | 16       | 
    5 | INT          | 5            | 20       | 
    6 | INT          | 0            | 24       | 
    7 | INT          | 16           | 28       | 
    8 | INT          | 160          | 32       | 
    9 | ARRAY        | a            | 36       | int[10][16]
   10 | INT          | 20           | 676      | 
   11 | ARRAY        | b            | 680      | float[20]
   12 | BLOCK        | main         | .L0      | Table 1, Params#: 0

Table 1:
  #   |   DataType   |     Name     |  Offset  |    Attr
------+--------------+--------------+----------+------------
    0 | INT          | (anonymous)  | 0        | 
    1 | INT          | m            | 4        | 
    2 | INT          | n            | 8        | 
    3 | INT          | 5            | 12       | 
    4 | ARRAY        | c            | 16       | bool[5]
    5 | BLOCK        | (anonymous)  |          | Table 2
    6 | BLOCK        | (anonymous)  |          | Table 3

Table 2:
  #   |   DataType   |     Name     |  Offset  |    Attr
------+--------------+--------------+----------+------------

Table 3:
  #   |   DataType   |     Name     |  Offset  |    Attr
------+--------------+--------------+----------+------------
    0 | BLOCK        | (anonymous)  |          | Table 4

Table 4:
  #   |   DataType   |     Name     |  Offset  |    Attr
------+--------------+--------------+----------+------------


Instruction sequence:
.L0    ( MOV,   1:1  ,   0:3  ,        )
       ( MOV,   1:2  ,   0:4  ,        )
       ( MOV,   1:5  ,   0:6  ,        )
       ( MUL,   1:5  ,   0:7  ,   1:6  )
       ( ADD,   1:6  ,   0:3  ,   1:6  )
       ( MOV,   1:5  ,   1:6  ,        )
       (MOVT,   0:9  ,   0:3  ,   1:5  )
       ( MOV,   1:5  ,   0:4  ,        )
       (MOVT,   0:11 ,   0:4  ,   1:5  )
       ( MOV,   1:5  ,   0:6  ,        )
       ( MUL,   1:5  ,   0:7  ,   1:6  )
       ( ADD,   1:6  ,   0:3  ,   1:6  )
       ( MOV,   1:5  ,   1:6  ,        )
       (MOVS,   1:7  ,   0:9  ,   1:5  )
       ( MOV,   1:6  ,   1:7  ,        )
       ( MOV,   1:5  ,   1:6  ,        )
       ( MOV,   1:6  ,   0:4  ,        )
       (MOVS,   1:6  ,   0:11 ,   1:6  )
       (MOVT,   0:11 ,   1:6  ,   1:5  )
       ( ADD,   1:1  ,   1:2  ,   1:5  )
       ( MOV,   1:6  ,   1:5  ,        )
       ( MOV,   1:5  ,   1:6  ,        )
       (MOVS,   1:5  ,   1:4  ,   1:5  )
       ( JNZ,   1:5  ,        , .L3    )
       ( JMP,        ,        , .L4    )
.L3    ( MOV,   1:1  ,   0:3  ,        )
       ( JMP,        ,        , .L5    )
.L4    ( MOV,   3:0  ,   1:1  ,        )
       (MOVS,   3:0  ,   1:4  ,   3:0  )
       ( JNZ,   3:0  ,        , .L1    )
       ( JMP,        ,        , .L5    )
.L1    ( MOV,   3:1  ,   1:2  ,        )
       (MOVS,   3:1  ,   1:4  ,   3:1  )
       ( JNZ,   3:1  ,        , .L2    )
       ( JMP,        ,        , .L5    )
.L2    ( MOV,   1:2  ,   0:4  ,        )
.L5    ( RET,        ,        ,        )
